src/core/public/keyboard_shortcut/keyboard_shortcut_service.ts

```diff
@@ -0,0 +1,129 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Any modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+import { KeyboardShortcutSetup, KeyboardShortcutStart, ShortcutDefinition } from './types';
+
+/**
+ * @internal
+ * @experimental
+ */
+export class KeyboardShortcutService {
+  private shortcuts = new Map<string, ShortcutDefinition[]>();
+
+  public setup(): KeyboardShortcutSetup {
+    return {
+      register: (shortcuts) => this.register(shortcuts),
+    };
+  }
+
+  public start(): KeyboardShortcutStart {
+    this.startEventListener();
+
+    return {
+      register: (shortcuts) => this.register(shortcuts),
+      unregister: (id) => this.unregister(id),
+    };
+  }
+
+  public stop() {
+    this.stopEventListener();
+    this.shortcuts.clear();
+  }
+
+  private normalizeKeyboardShortcutString = (str: string): string => str.toLowerCase();
+
+  private getNamespacedIdForKeyboardShortcut = (
+    shortcut: Pick<ShortcutDefinition, 'id' | 'pluginId'>
+  ) =>
+    `${this.normalizeKeyboardShortcutString(shortcut.id)}.${this.normalizeKeyboardShortcutString(
+      shortcut.pluginId
+    )}`;
+
+  private getEventKeyString = (event: KeyboardEvent): string => {
```

Comment:
how about `getNormalizedKeyFromKeyboardEvent`.

But also, i see why we need two functions to normalize... One we need to normalize for developeres who are registering a keyboard shortcut, and one we need to normalize for the keyboard event. Is there a good way to reduce code duplication between the two?